sdk/client/src/verifying_client.rs

```diff
@@ -0,0 +1,516 @@
+        Ok(Response::new(response, state.clone()))
+    }
+}
+
+impl From<MethodRequest> for VerifyingRequest {
+    fn from(request: MethodRequest) -> Self {
+        match request {
+            MethodRequest::GetAccount((address,)) => verifying_get_account(address),
+            _ => todo!(),
+        }
+    }
+}
+
+// TODO(philiphayes): add separate type for each MethodRequest? like:
+//
+// ```
+// struct GetAccount((address,));
+//
+// enum MethodRequest {
+//     GetAccount(GetAccount),
+//     // ..
+// }
+// ```
+//
+// would allow the from(MethodRequest) above to call a method on the enum inner
+// instead of these ad-hoc methods i think
+
+fn verifying_get_account(address: AccountAddress) -> VerifyingRequest {
+    let request = MethodRequest::GetAccount((address,));
+    let subrequests = vec![
+        MethodRequest::GetAccountStateWithProof(diem_root_address(), None, None),
+        MethodRequest::GetAccountStateWithProof(address, None, None),
+    ];
+    let callback: RequestCallback = |ctxt, subresponses| match subresponses {
+        [MethodResponse::GetAccountStateWithProof(ref diem_root), MethodResponse::GetAccountStateWithProof(ref account)] =>
+        {
+            let diem_root_with_proof: AccountStateWithProof = diem_root.try_into()?;
+            let account_state_with_proof: AccountStateWithProof = account.try_into()?;
+
+            let latest_li = ctxt.state_proof.0.ledger_info();
+            let state_version = latest_li.version();
+
+            let address = match ctxt.request {
+                MethodRequest::GetAccount((address,)) => *address,
+                _ => panic!("should not happen"),
+            };
+
+            diem_root_with_proof.verify(latest_li, state_version, diem_root_address())?;
+            account_state_with_proof.verify(latest_li, state_version, address)?;
+
+            // TODO(philiphayes): it seems wasteful to lookup the whole diem_root
+            // account, verify the proofs, etc... just to get a list of supported
+            // currency codes. Would it make sense for the AccountView to just
+            // list _all_ BalanceResource's under balance? Then we could avoid
+            // this extra lookup. It's even more problematic when a batch
+            // includes many GetAccount requests, as this batch handling logic
+            // isn't (currently) smart enough to dedup the diem_root lookups.
+            let diem_root_blob = diem_root_with_proof
+                .blob
+                .ok_or_else(|| format_err!("missing diem_root accound somehow"))?;
+            let diem_root = AccountState::try_from(&diem_root_blob)?;
+            let currencies: Option<RegisteredCurrencies> = diem_root.get_config()?;
+            let currency_codes = currencies
+                .as_ref()
+                .map(RegisteredCurrencies::currency_codes)
+                .ok_or_else(|| format_err!("diem_root missing registered currencies"))?;
```

Comment:
cc @tzakian @sblackshear

Do you think it would make sense to list all `BalanceResource`s when displaying an `AccountView`, rather than only currently registered currencies in `diem_root`? Or all currency-like `BalanceResource`s? It would make these verifying `get_account` calls 2x cheaper/simpler.

I'm mostly wondering if an account would ever contain unregistered currencies. Thanks : )

Reply:
An account should never hold a balance in an unregistered currency. So, using the set of registered currencies to look for balances under the account would work, or finding all BalanceResources would be the same.

Reply:
I'll see if I can address this in a follow up PR